oioioi/forum/models.py

```diff
@@ -192,22 +193,26 @@ class Post(models.Model):
     class PostsWithReactionsSummaryManager(models.Manager):
         def get_queryset(self):
             qs = super(Post.PostsWithReactionsSummaryManager, self).get_queryset()
-            for field_name, rtype in [
-                ('upvotes_count', 'UPVOTE'),
-                ('downvotes_count', 'DOWNVOTE'),
-            ]:
-                # In Django >=2.0 it can can be simplified with Count(filter=Q(...))
+
+            for rtype, attr_name in POST_REACTION_TO_COUNT_ATTR.items():
                 reaction_count_agg = {
-                    field_name: models.Sum(
-                        models.Case(
-                            models.When(reactions__type_of_reaction=rtype, then=1),
-                            default=0,
-                            output_field=models.IntegerField(),
-                        )
+                    attr_name: models.Count(
+                        'reactions', 
+                        filter=models.Q(reactions__type_of_reaction=rtype)
                     )
                 }
                 qs = qs.annotate(**reaction_count_agg)
 
+            max_count = getattr(settings, 'FORUM_REACTIONS_TO_DISPLAY', 10)
+            for rtype, attr_name in POST_REACTION_TO_PREFETCH_ATTR.items():
+                qs = qs.prefetch_related(
+                    models.Prefetch(
+                        'reactions', 
+                        to_attr=attr_name,
+                        queryset=PostReaction.objects.filter(type_of_reaction=rtype).select_related('author')[:max_count], 
```

Comment:
it would be nice to add some ordering here, otherwise we can get random results (especially important during tests): i.e. `order_by('-pk')`

also you can add Meta class to PostReaction like:
```
class Meta:
        ordering = ['-pk']
```